### ISSUE TITLE
Question regarding your talk at eXchange 2016

### ISSUE THREAD

**Issue by clojj** [ndmitchell/spaceleak#1]
Hi, great talk! 

Regarding slide 4 of your talk, it is not clear to me, what you mean by "ideal solution".
Isn't [1..10] or [1,2,3,4,5,6,7,8,9,10] as an argument actually producing an graph-of-expressions ?

So actually isn't Haskell also laying-out this list first completely... just not as a list of Integers but as a list of thunks ?

My first reaction to this slide was, wow here is a language 'producing' Integers on the fly, or by demand.. if you will.
But this actually isn't really the case, right ?

Best,
Joerg

**Comment by ndmitchell** [comment]
Haskell lays out a single think, literally 1..10, taking 1 memory cell. As you evaluate it produces 1:2..10, now taking 3 memory cells, but that happens only when 1 is required. As a result Haskell really does produce the integers on the fly.

**Comment by clojj** [comment]
Ok.
So the construct [1..n] is semantically different from , say
[1,4,6,7,8,100,200,300] ?
[1..10] will create 'on demand' vs. [1,4,6,7,8,100,200,300] must be laid
out as some kind of thunks with constants, right ?

2016-10-07 11:49 GMT+02:00 Neil Mitchell notifications@github.com:

**Comment by ndmitchell** [comment]
Correct. [1..10] is equivalent to enumFromTo 1 10, so it's a function that is evaluated lazily.
